### PR TITLE
DEMO: Upcast bug with Dexter

### PR DESCRIPTION
**Pull request by paulcsmith**
To run this:

* Clone this repo and check out this branch
* Run `scripts/setup` to setup the Docker containers
* Run `docker-compose run --rm app crystal spec`

You'll see:

```
BUG: trying to upcast Dexter::Formatters::JsonLogFormatter <- Dexter::Formatters::BaseLogFormatter (Exception)
   from ???
   from ???
   from ???
```

### CODE REVIEW

File: config/database.cr [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
 
+# Adding this struct makes everything work
+#
+# Removing it causes:
+#
+# BUG: trying to upcast Dexter::Formatters::JsonLogFormatter <- Dexter::Formatters::BaseLogFormatter (Exception)
+#   from ???
+#   from ???
+#   from ???
+struct NotEvenUsedAnywhere < Dexter::Formatters::BaseLogFormatter
```
Comment (paulcsmith) [inline config/database.cr]: This fixes the issue...when it is commented out the compiler bug pops up

File: config/database.cr [hunk 2]
```diff
@@ -1,5 +1,19 @@
 database_name = "avram_dev"
 
+# Adding this struct makes everything work
+#
+# Removing it causes:
+#
+# BUG: trying to upcast Dexter::Formatters::JsonLogFormatter <- Dexter::Formatters::BaseLogFormatter (Exception)
```
Comment (paulcsmith) [inline config/database.cr]: Link to JsonLogFormatter: https://github.com/luckyframework/dexter/blob/master/src/dexter/formatters/json_log_formatter.cr

**Comment by jwoertink** [comment]
I'm trying to verify if this is fixed with crystal 0.29, but I get `(DB::ConnectionRefused)` running the `docker-compose run --rm app crystal spec`, so I'm not sure. But I can run `./scripts/test` on this branch, and all specs pass with crystal 0.29....

File: config/database.cr (hunk 1 above)
Comment (paulcsmith) [reply config/database.cr]: @jwoertink Did you comment this out when testing?

Comment (jwoertink) [reply config/database.cr]: nope. Didn't comment out anything, but I did have to update the repo for crystal 0.29 support. (i.e. fix requires)

Comment (paulcsmith) [reply config/database.cr]: I believe you need to comment this struct out for the bug to pop up.

Comment (jwoertink) [reply config/database.cr]: oh, I read it wrong lol. Ok

Comment (jwoertink) [reply config/database.cr]: Alright, commented that struct out, and `./scripts/test` still runs fine. Also:

```
[08:27AM] avram ((619d94f...))$ docker-compose run --rm app crystal spec/
Starting avram_db_1 ... done
.........................................................................*............*.....................................................................................................................................................................................................

Pending:
  Avram::Form .create on failure logs the failure if a logger is set
  Avram::Form .update on failure logs the failure

Finished in 17.41 seconds
284 examples, 0 failures, 0 errors, 2 pending
```

Comment (paulcsmith) [reply config/database.cr]: Oh awesome! I'll give. it a shot too and close this out 🤞

**Comment by paulcsmith** [comment]
Yes!! Looks like it is indeed working now. Thanks for checking this @jwoertink :D

**Comment by paulcsmith** [comment]
More details here: https://github.com/crystal-lang/crystal/pull/7787

**Pull request closed without merging**